metricbeat/beater/module.go

```diff
@@ -0,0 +1,204 @@
+}
+
+// newModuleWrappers creates new Modules and their associated MetricSets based
+// on the given configuration. It constructs the supporting filters and
+// publisher client and stores it all in a moduleWrapper.
+func newModuleWrappers(
+	modulesConfig []*common.Config,
+	r *mb.Register,
+	publisher *publisher.Publisher,
+) ([]*moduleWrapper, error) {
+	modules, err := mb.NewModules(modulesConfig, r)
+	if err != nil {
+		return nil, err
+	}
+
+	// Wrap the Modules and MetricSet's.
+	var wrappers []*moduleWrapper
+	var errs multierror.Errors
+	for k, v := range modules {
+		debugf("initializing Module type %s, %T=%+v", k.Name(), k, k)
+		f, err := filter.New(k.Config().Filters)
+		if err != nil {
+			errs = append(errs, errors.Wrapf(err, "module %s", k.Name()))
+			continue
+		}
+
+		mw := &moduleWrapper{
+			Module:    k,
+			filters:   f,
+			pubClient: publisher.Connect(),
+		}
+		wrappers = append(wrappers, mw)
+
+		msws := make([]*metricSetWrapper, 0, len(v))
+		for _, ms := range v {
+			debugf("initializing MetricSet type %s/%s, %T=%+v",
+				ms.Module().Name(), ms.Name(), ms, ms)
+			msw := &metricSetWrapper{
+				MetricSet: ms,
+				parent:    mw,
+				stats:     new(expvar.Map).Init(),
+			}
+			msws = append(msws, msw)
+
+			// Initialize expvar stats for this MetricSet.
+			fetches.Set(fmt.Sprintf("%s-%s", mw.Name(), msw.Name()), msw.stats)
+			msw.stats.Add(successesKey, 0)
+			msw.stats.Add(failuresKey, 0)
+			msw.stats.Add(eventsKey, 0)
+		}
+		mw.children = msws
+	}
+
+	return wrappers, errs.Err()
+}
+
+// metricSetWrapper methods
+
+// startFetching performs an immediate fetch for the specified host then it
+// begins continuous timer scheduled loop to fetch data. To stop the loop the
+// done channel should be closed. On exit the method will decrement the
+// WaitGroup counter.
+//
+// startFetching manages fetching for a single host so it should be called once
+// per host.
+func (msw *metricSetWrapper) startFetching(
+	done <-chan struct{},
+	wg *sync.WaitGroup,
+	host string,
+) {
+	defer wg.Done()
+
+	// Fetch immediately.
+	err := msw.fetch(host)
+	if err != nil {
+		logp.Err("fetch error: %v", err)
+	}
+
+	// Start timer for future fetches.
+	t := time.NewTicker(msw.Module().Config().Period)
+	defer t.Stop()
+	for {
+		select {
+		case <-done:
+			return
+		case <-t.C:
+			err := msw.fetch(host)
+			if err != nil {
+				logp.Err("%v", err)
+			}
+		}
+	}
+}
+
+// fetch invokes the appropriate Fetch method for the MetricSet and publishes
+// the result using the publisher client. This method will recover from panics
+// and log a stack track if one occurs.
+func (msw *metricSetWrapper) fetch(host string) error {
+	defer logp.Recover(fmt.Sprintf("recovered from panic while fetching "+
+		"'%s/%s' for host '%s'", msw.parent.Name(), msw.Name(), host))
+
+	switch fetcher := msw.MetricSet.(type) {
+	case mb.SingleEventFetcher:
+		return msw.singleEventFetch(fetcher, host)
+	case mb.MultiEventFetcher:
```

Comment:
Could we call this `EventFetcher` and `EventsFetcher` or `EventListFetcher`. I'm a little bit worried that MultiEvent could lead to a wrong understanding. In the case of processes, I see all the events/documents together as one event which was catched even if we do not store it as nested document. All the documents have the same timestamp and identifier.